### ISSUE TITLE
Test Weave Net with k8s

### ISSUE THREAD

**Issue by nonsense** [ipfs/testground#318]
https://github.com/weaveworks/weave

**Comment by nonsense** [comment]
https://github.com/weaveworks/weave/blob/master/site/tasks/manage/dynamically-attach-containers.md - looks like Weave might provide the functionality we need for `sidecar` to be able to dynamically assign IPs to containers after they are already running. cc @Stebalien

**Comment by Stebalien** [comment]
Nice! That does look like it will work.

**Comment by nonsense** [comment]
I got `flannel` (secondary) and `weave` (primary k8s network) to work, but I am trying the other way around - `flannel` for a primary, and `weave` for a secondary, as we have much more control over `weave`, and it'd make our lives easier to use it for the data plane. Unfortunately there seems to be some issue when running it in that setup, so still have to figure out what the configuration problem is.

**Comment by nonsense** [comment]
A short update on this one - for some reason I can't get `weave` to work as a secondary network alongside any other network. At the moment `weave` must always be the primary network, and it is not clear why. It is possible that this is some weird bug between Multus CNI and weave, so I will try to run `weave` as a secondary network via https://github.com/cni-genie/CNI-Genie , which is another CNI similar to Multus, which allows for multiple network interfaces to be attached to pods.

**Comment by nonsense** [comment]
Current setup can be found at: https://github.com/ipfs/testground/pull/323

**Comment by nonsense** [comment]
Worth mentioning that both Multus CNI and CNI-Genie CNI (the meta CNIs that provide support for multiple CNIs on k8s) fail with `weave` as a second network for some reason.

Closing this for now, as `weave` is mostly working on k8s, apart from parts of its functionality (and it is not clear if these parts are designed to be used in k8s). I will probably be coming back to this, and try to get the `Dynamically Attaching and Detaching Applications` part fully working, but for now will move on evaluating other CNIs.

**Comment by nonsense** [comment]
After my call with @raulk yesterday, and noticing that when adding `weave` as a second network, the IPs added to `ipset` by `weave` are actually the `flannel` IPs (the default network, and the pod ID), I looked up the code, and it is pretty clear from `github.com/weaveworks/weave/npc/controller.go` that weave is always picking up the `PodIP` when adding entries to `ipset` (which then are used by `iptables`). This explains why the second `weave` network never worked.

I will continue to look into it, but also review `flannel` back again, but it seems like we won't be able to use `weave` as a second data plane in the format I had in mind - `flannel` as a default k8s network, and `weave` as a dynamically attached/detached network :(

Ultimately it looks like dynamically attaching/detaching network interfaces might not be supported (also see https://github.com/containernetworking/cni#what-might-cni-do-in-the-future - not sure if they speak about the actual networks parameters, or pods attaching to existing networks).

I continue to investigate how we can dynamically attach and detach from networks at runtime, in the meantime:

1. One solution to our `multi-network problem` that comes to mind is - create N flannel networks (for example 3 or 5), start a test plan with M pods across many hosts, that attach to all the N networks at start, and have sidecar just `ifconfig down` interfaces that a given test plan instance is not supposed to use.

2. Another solution that comes to mind is defining the topology at start-up time and just attaching to the correct networks on pod creation.

cc @Stebalien

**Comment by nonsense** [comment]
I got lucky an my question at the CNI Slack got answers. `Bryan Boreham` helped out, and when explained the problem, suggested we turn off `weave-npc` and start `weave net` container with `EXPECT_NPC=0`, which does the trick. Basically `weave-npc` (Network Policy Controller) was blocking our connections previously.

`weave attach` now works as well!

**Comment by raulk** [comment]
@nonsense that is fantastic! So are we capable of attaching a container to multiple subnets nested inside the default net? And do pings work across containers in different hosts that are attached to the same subnet? And are subnets isolated, i.e. two containers connected to different subnets cannot reach each other?

**Comment by nonsense** [comment]
> So are we capable of attaching a container to multiple subnets nested inside the default net?

We are capable of attaching a container to `one` subnet nested inside the default net (or just to the default net). It seems that we `cannot` have multiple interfaces (2 or more) for `weave`. We have attachments to two networks in total - one to `flannel`, and one to `weave` (where we have control over which subnet).

> And do pings work across containers in different hosts that are attached to the same subnet?

Yes, they do. The subnet is split between the hosts, because `weave` is managing it dynamically, so if you use 10.33.10.0/24 , you will most probably initially see your 2 pods end up at 10.33.10.1 and 10.33.10.128, because the subnet is split in half, between the 2 hosts initially.

> And are subnets isolated, i.e. two containers connected to different subnets cannot reach each other?

By default (when not running `weave-npc` (network policy controller)) they are not isolated - there is nothing stopping 10.33.10.128 to communicate with 10.32.0.2 for example. We would have to add `iptable` rules to block this traffic - this is what the `weave-npc` is for, but it doesn't work for our purpose, because we use `weave` as a secondary network, and not the primary k8s pod network.

**Comment by nonsense** [comment]
@raulk an update on the first question: `So are we capable of attaching a container to multiple subnets nested inside the default net?`

It is possible to attach multiple subnets (from the default net) to a container, but there is a caveat: https://github.com/weaveworks/weave/pull/3343 - weave currently maps interface name to the veth on the host, so unless we fork weave (or get this PR merged upstream), it is not working, as there is a conflict.

I tested with a forked version and the commit from above - https://github.com/nonsense/weave/commit/c850c2f863cd4e42e7d83241b26e3ed66bd8845e - and it works - see screenshot.

On the screenshot you see 2 pods on different hosts, each of which attached to 4 interfaces (2 weave, 1 flannel (eth0), 1 lo), where `net1` and `net2` are `weave`, and both interfaces are reachable.

If we want to isolate the subnets, we'd have to drop the `default` route... currently `ip route` returns:

```
root@samplepod3:/# ip route
default via 100.96.1.1 dev eth0
10.32.0.1 dev net1 scope link
10.35.80.0/24 dev net1 proto kernel scope link src 10.35.80.1
10.36.79.0/24 dev net2 proto kernel scope link src 10.36.79.1
100.96.0.0/16 via 100.96.1.1 dev eth0
100.96.1.0/24 dev eth0 proto kernel scope link src 100.96.1.2
```

---

Bottom line: it looks like this setup is good for our usecase, so I continue with trying to test it with `sidecar`, for which I have to modify `sidecar` to detect these networks, as it currently works only with Docker networks.
